cpp/src/arrow/array/data.h

```diff
@@ -564,4 +565,128 @@ Result<std::shared_ptr<ArrayData>> GetArrayView(const std::shared_ptr<ArrayData>
                                                 const std::shared_ptr<DataType>& type);
 
 }  // namespace internal
+
+struct ARROW_EXPORT ConcreteArraySpan {
```

Comment:
@felipecrv What do you think about these API additions? They look like a reasonable addition me.

Reply:
I have concerns about adding a new class hierarchy to represent arrays, even though I understand why these help here. Can it be moved to an `_internal.h` header before we settle on the naming and design? cc @js8544